lua/plenary/enum.lua

```diff
@@ -0,0 +1,83 @@
+local Enum = {}
+
+local function make_enum(tbl)
+  local enum = {}
+
+  local Variant = {}
+  Variant.__index = Variant
+
+  local function newVariant(i)
+    return setmetatable({_id = i}, Variant)
+  end
+
+  -- we don't need __eq because the __eq metamethod will only ever be invoked when they both have the same metatable
+
+  function Variant:__lt(o)
+    return self._id < o._id
+  end
+
+  function Variant:__gt(o)
+    return self._id > o._id
+  end
+
+  function Variant:__tostring()
+    return tostring(self._id)
+  end
+
+  function Variant:id()
+    return self._id
+  end
+
+  local function find_next_idx(enum, i)
+    while true do
+      if not enum[i] then return i end
```

Comment:
I am wondering if we should just error instead of trying to find the next available index like some other languages do

Reply:
I think it's best to error out, since, if the user is defining the enum like this, they are likely doing it out of order, which would obscure out the actual values of the members

Reply:
I agree that we should error out. I just checked in rust and it will also error out.